### PR TITLE
Add supervisory parens for bitwise operations

### LINKED ISSUE #2424: Bit shift operators are not parenthesised as expected
```js
a << b >>> c;
```

Should become:

```js
(a << b) >>> c;
```

https://prettier.io/playground/#%7B%22content%22%3A%22%5CnA%3C%3Cb%3E%3E%3Ec%3B%5Cn%22%2C%22options%22%3A%7B%22printWidth%22%3A80%2C%22tabWidth%22%3A2%2C%22singleQuote%22%3Afalse%2C%22trailingComma%22%3A%22none%22%2C%22bracketSpacing%22%3Atrue%2C%22jsxBracketSameLine%22%3Afalse%2C%22parser%22%3A%22babylon%22%2C%22semi%22%3Atrue%2C%22useTabs%22%3Afalse%2C%22doc%22%3Afalse%2C%22ast%22%3Afalse%7D%7D

### PR DESCRIPTION
**Pull request by azz**
Adds parens for all of these cases:

* Any combination of bitshift operators, e.g:
    ```js
    x << y >> z;
    x >>> y << z;
    x >>> y >>> z;
   ```
* LHS parens for the other bitwise operators:
    ```js
    x ^ y ^ z;
    x & y & z;
    x | y | z;
    ```

Fixes #2424

### CODE REVIEW

File: tests/binary_math/__snapshots__/jsfmt.spec.js.snap [hunk 1]
```diff
@@ -17,9 +17,9 @@ var from = offset > left ? 0 : (left - offset) >> level;
 var to = ((right - offset) >> level) + 1;
 if (rawIndex < 1 << (list._level + SHIFT)) {
 }
-var res = size < SIZE ? 0 : (size - 1) >>> SHIFT << SHIFT;
+var res = size < SIZE ? 0 : ((size - 1) >>> SHIFT) << SHIFT;
 sign = 1 - 2 * (b[3] >> 7);
 exponent = (((b[3] << 1) & 0xff) | (b[2] >> 7)) - 127;
-mantissa = ((b[2] & 0x7f) << 16) | (b[1] << 8) | b[0];
+mantissa = (((b[2] & 0x7f) << 16) | (b[1] << 8)) | b[0];
```
Comment (azz) [inline tests/binary_math/__snapshots__/jsfmt.spec.js.snap]: This one looks a _little_ excessive but it is adding parens from `x | y | z` to `(x | y) | z`.

**Comment by azz** [comment]
Realised we never want to add parens for `x | y | z` due to the common bitflag pattern.

**Comment by bakkot** [comment]
It's not just `|` - I don't think we should be adding parens for `&` or `^` either (that is, e.g. `a & b & c`), for the same reason we don't add them to `a + b + c`: those operation are associative, so it doesn't matter at all how they're parenthesized. And when it's just a single operation, there's no risk of confusing people who haven't memorized the precedence rules. If anything I think prettier adding the parentheses runs the risk of confusing people into thinking they aren't associative.

(I can open an issue for this if you like.)

**Comment by azz** [comment]
Go for it.

**Pull request merged**